### ISSUE TITLE
Figure out why loading big pages is so slow

### ISSUE THREAD

**Issue by tangjeff0** [athensresearch/athens#240]
My guess is that it is `any-char` in the parser. Please see the file attached.

I timed three different functions against one another.

1. `identity` - 0.14-0.295 msecs
1. `parse-and-render` - 12681-14405 msecs
1. `parser/block-parser` 14813-15369 msecs

I ran each function 5 times. The first four times have no print logs. The last iteration prints the output of the function.

`cmd-f` "Elapsed time".

[tmp.log](https://github.com/athensresearch/athens/files/4922329/tmp.log)

```

(defn time-me
  [data f]
  (doall
    (for [x data]
      (f x)))
  nil)


(map (fn [f]
       (dotimes [i 5]
         (if (= i 4)
           (time (time-me data (comp prn f)))
           (time (time-me data f)))))
  [identity parse-and-render parser/block-parser])
```

**Comment by roryokane** [comment]
It makes sense that `any-char` would be slow, so that’s probably a good guess.

The parse rule introduced [in commit `9433e79`](https://github.com/athensresearch/athens/commit/9433e79d2c30f5e7ab2ccd9feb953132b0e30b96#diff-881af7a0c0560f5d6837f19e442a8132R20) demonstrates a more efficient way to parse “any chars”:

~~~
<any-non-page-link-chars> = #'[^\\[\\]]*'
~~~

We can probably replace other uses of `any-chars` with similar constructions.

I didn’t write the parser with regexes like that to start with because such regexes slightly increase the difficulty of adding new syntax that may be nested. For example, if page links could contain `**bold**` text, the above regex would have to be `#'([^*\\[\\]]|\\*[^*])*'` (or you could use a shorter regex paired with another parse rule). The information that the syntax for bold text is `**` would be duplicated both in the `bold` rule and in regexes of rules that can contain bold text.

**Comment by nthd3gr33** [comment]
Thank you for looking into this issue @roryokane 🙏

**Comment by HaojiXu** [comment]
I read the issues section of instaparse, and here's their advice: 
* <https://github.com/Engelberg/instaparse/issues/131>
* <https://github.com/Engelberg/instaparse/blob/master/docs/Performance.md>

Because people over there mentioned avoid regex combinations & use LL(1)-like rules as much as possible, so I made a experimental parser with their advices here: https://github.com/athensresearch/athens/compare/master...HaojiXu:master

I'm currently getting 1 second loading times when it comes to large pages, and also documented the parser with comments regarding expanding it.

One minor problem though is that Unicode regex matches didn't seem to work anymore - instead of matching letters, marks, and numbers (for characters in these Unicode classes) it only matches the three literal characters (L,M,N). So for now I changed it to a ASCII-only hashtag parser, and I'm interested to know how can I solve it.

Thanks for the great contributions!
